parsl/app/futures.py

```diff
@@ -31,15 +34,26 @@ def parent_callback(self, parent_fu):
         Returns:
             - None
         """
-
         e = parent_fu._exception
         if e:
             self.set_exception(e)
         else:
             self.set_result(self.file_obj)
+            if self.file_obj.scheme == 'file' and os.path.isfile(self.file_obj.filepath):
+                if not self.file_obj.timestamp:
+                    self.file_obj.timestamp = datetime.fromtimestamp(stat(self.file_obj.filepath).st_ctime, tz=timezone.utc)
+                if not self.file_obj.size:
+                    self.file_obj.size = stat(self.file_obj.filepath).st_size
+                if not self.file_obj.md5sum:
+                    self.file_obj.md5sum = md5(open(self.file_obj, 'rb').read()).hexdigest()
+            else:
+                self.file_obj.timestamp = datetime.now(tz=timezone.utc)
+            if self.data_flow_kernel:
+                self.data_flow_kernel.register_as_output(self.file_obj, self.app_fut.task_record)
 
     @typeguard.typechecked
-    def __init__(self, fut: Future, file_obj: File, tid: Optional[int] = None) -> None:
+    def __init__(self, fut: Future, file_obj: File, tid: Optional[int] = None, app_fut: Optional[Future] = None,
+                 dfk: Optional[Any] = None) -> None:
```

Comment:
Maybe it makes sense for this dfk parameter to be compulsory, rather than `Optional`: I think every use of DataFuture in the codebase can always be associated with a DFK. In this PR, there's only one place that `dfk` is not specified, and I think that is straightforward to fix up.

Reply:
Sounds good to me. I will work on the fixes, I assume that I should re-order the keyword arguments so that the dfk comes before any optional keywords.

Reply:
If its being used as a keyword argument (which is a good thing), I don't think it really matters about the order.

You can *force* people to use `dfk`, `tid` and `app_fut` as kwargs and never as positional args (which is also a good thing, I think) with a `*` like this:

```
def __init__(self, fut: Future, file_obj: File, *, tid: Optional[int] = None, app_fut: Optional[Future] = None, dfk: Optional[Any] = None) 
```

Recently I've mostly been pushing on forcing keyword args to be used as keyword args, rather than caring about the specific order, because I think that referencing things by name when there are many args is easier to read later on.